// @ramp-network/react-native-sdk/lib/commonjs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "WidgetEventTypes", {
  enumerable: true,
  get: function () {
    return _types.WidgetEventTypes;
  }
});
exports.default = void 0;

var _reactNative = require("react-native");

var _types = require("./types");

var _utils = require("./utils");

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const RampSdkNativeModule = _reactNative.NativeModules.RampSdk;
const RampEvents = new _reactNative.NativeEventEmitter(RampSdkNativeModule);

class RampSdk {
  constructor(config) {
    this.config = config;

    _defineProperty(this, "_listeners", (0, _utils.initEventListenersDict)());

    _defineProperty(this, "_instanceId", void 0);

    this._instanceId = (0, _utils.getRandomIntString)();

    this._subscribeToRampEvents();
  }

  show() {
    RampSdkNativeModule.runRamp({
      instanceId: this._instanceId,
      ...this.config
    });
    return this;
  }

  on(type, callback) {
    if (type !== '*' && !this._listeners[type]) {
      // tslint:disable-next-line:no-console
      console.warn(`Unknown / unsupported event name - '${type}'. This listener will have no effect.`);
    }

    if (type === '*') {
      const allTypes = Object.values(this._listeners);
      allTypes.forEach(eventHandlers => eventHandlers.push(callback));
    } else {
      this._listeners[type].push(callback);
    }

    return this;
  }

  unsubscribe(type, callback) {
    if (type === '*') {
      const allTypes = Object.entries(this._listeners);
      allTypes.forEach(([key, eventHandlers]) => {
        const filteredHandlers = eventHandlers.filter(c => c !== callback);
        this._listeners[key] = filteredHandlers;
      });
    } else {
      this._listeners[type] = this._listeners[type].filter(c => c !== callback);
    }

    return this;
  }

  _subscribeToRampEvents() {
    RampEvents.addListener('onRamp', event => {
      console.log('onRamp', event);

      if (event.instanceId !== this._instanceId) {
        return;
      }

      this._dispatchEvent({
        type: _types.WidgetEventTypes.PURCHASE_CREATED,
        payload: {
          purchase: event.purchase,
          purchaseViewToken: event.purchaseViewToken,
          apiUrl: event.apiUrl
        }
      });
    });
    RampEvents.addListener('onRampDidClose', event => {
      console.log('onRampDidClose', event);

      if (event.instanceId !== this._instanceId) {
        return;
      }

      this._dispatchEvent({
        type: _types.WidgetEventTypes.WIDGET_CLOSE,
        payload: null
      });
    });
  }

  _dispatchEvent(event) {
    const {
      type
    } = event;

    this._listeners[type].forEach(callback => callback(event));
  }

}

exports.default = RampSdk;
